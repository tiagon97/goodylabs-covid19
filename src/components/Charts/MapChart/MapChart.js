import React, { memo } from 'react';
import { ComposableMap, Geographies, Geography } from 'react-simple-maps';
import PropTypes from 'prop-types';
import Url from '../../../constants';

const MapChart = ({ setTooltipContent, countriesData }) => (
  <>
    <ComposableMap height={420} data-tip="" projectionConfig={{ scale: 150 }}>
      <Geographies geography={Url.geoUrl}>
        {({ geographies }) =>
          geographies.map((geo) => (
            <Geography
              key={geo.rsmKey}
              geography={geo}
              onMouseEnter={() => {
                const { NAME, ISO_A2 } = geo.properties;
                const result = countriesData.filter((country) => country.CountryCode === ISO_A2);
                if (!result.length) {
                  setTooltipContent(`${NAME}\nno data available`);
                } else {
                  const MultiLineString = `Total cases: ${result[0].TotalConfirmed}\nTotal deaths: ${result[0].TotalDeaths}\nTotal recovered: ${result[0].TotalRecovered}`;
                  setTooltipContent(`${NAME}\n${MultiLineString} `);
                }
              }}
              onMouseLeave={() => {
                setTooltipContent('');
              }}
              style={{
                default: {
                  fill: '#D6D6DA',
                  outline: 'none',
                },
                hover: {
                  fill: '#F53',
                  outline: 'none',
                },
                pressed: {
                  fill: '#E42',
                  outline: 'none',
                },
              }}
            />
          ))
        }
      </Geographies>
    </ComposableMap>
  </>
);
MapChart.propTypes = {
  setTooltipContent: PropTypes.func,
  countriesData: PropTypes.instanceOf(Array).isRequired,
};

MapChart.defaultProps = {
  setTooltipContent: () => 'no data',
};

export default memo(MapChart);
