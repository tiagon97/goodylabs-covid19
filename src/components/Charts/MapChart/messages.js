const messages = {
  noData: 'No data available',
  totalCases: 'Total cases: ',
  totalDeaths: 'Total deaths: ',
  totalRecovered: 'Total recovered: ',
};

export default messages;
