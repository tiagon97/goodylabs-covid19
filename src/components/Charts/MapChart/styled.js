import styled from 'styled-components';

import { Colors } from '../../../constants';

export const Header = styled.h4`
  align-self: center;
  margin-bottom: 8px;
  color: ${Colors.white};
`;

export const TooltipContainer = styled.div`
  display: flex;
  flex-direction: column;
  padding: 5px;
`;

export const DataContainer = styled.div``;

export const ValuesContainer = styled(TooltipContainer)`
  padding: 0;
`;
