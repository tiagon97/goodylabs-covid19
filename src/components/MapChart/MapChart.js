import React, { memo } from 'react';
import { ComposableMap, Geographies, Geography } from 'react-simple-maps';
import PropTypes from 'prop-types';

const geoUrl =
  'https://raw.githubusercontent.com/zcreativelabs/react-simple-maps/master/topojson-maps/world-110m.json';

const MultiLineString = `\n\nTotal cases: 23342 
Total deaths: 823 
Total recovered: 2122`;

const MapChart = ({ setTooltipContent }) => (
  <>
    <ComposableMap height={450} data-tip="" projectionConfig={{ scale: 150 }}>
      <Geographies geography={geoUrl}>
        {({ geographies }) =>
          geographies.map((geo) => (
            <Geography
              key={geo.rsmKey}
              geography={geo}
              onMouseEnter={() => {
                const { NAME } = geo.properties;
                setTooltipContent(`${NAME} ${MultiLineString}`);
              }}
              onMouseLeave={() => {
                setTooltipContent('');
              }}
              style={{
                default: {
                  fill: '#D6D6DA',
                  outline: 'none',
                },
                hover: {
                  fill: '#F53',
                  outline: 'none',
                },
                pressed: {
                  fill: '#E42',
                  outline: 'none',
                },
              }}
            />
          ))
        }
      </Geographies>
    </ComposableMap>
  </>
);
MapChart.propTypes = {
  setTooltipContent: PropTypes.func,
};

MapChart.defaultProps = {
  setTooltipContent: () => 'no data',
};

export default memo(MapChart);
