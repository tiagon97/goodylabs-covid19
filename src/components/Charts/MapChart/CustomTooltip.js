import React from 'react';
import PropTypes from 'prop-types';

import formatValue from '../../../utils';
import messages from './messages';
import { Header, TooltipContainer, DataContainer, ValuesContainer } from './styled';
import { StyledSpan } from '../../GlobalStatistics/styled';

const CustomTooltip = ({ header, data }) => (
  <TooltipContainer>
    <Header>{header}</Header>
    <DataContainer>
      {data ? (
        <ValuesContainer>
          <StyledSpan color="primary">{`${messages.totalCases}${formatValue`${data.totalConfirmed}`}`}</StyledSpan>
          <StyledSpan color="secondary">{`${messages.totalDeaths}${formatValue`${data.totalDeaths}`}`}</StyledSpan>
          <StyledSpan>{`${messages.totalRecovered}${formatValue`${data.totalRecovered}`}`}</StyledSpan>
        </ValuesContainer>
      ) : (
        messages.noData
      )}
    </DataContainer>
  </TooltipContainer>
);

CustomTooltip.propTypes = {
  header: PropTypes.string.isRequired,
  data: PropTypes.objectOf(PropTypes.number),
};

CustomTooltip.defaultProps = {
  data: null,
};

export default CustomTooltip;
