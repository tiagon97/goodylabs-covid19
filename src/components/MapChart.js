import React, { memo } from 'react';
import { ZoomableGroup, ComposableMap, Geographies, Geography } from 'react-simple-maps';
import styled from 'styled-components';

const geoUrl =
  'https://raw.githubusercontent.com/zcreativelabs/react-simple-maps/master/topojson-maps/world-110m.json';

const rounded = (num) => {
  if (num > 1000000000) {
    return `${Math.round(num / 100000000) / 10}Bn`;
  }
  if (num > 1000000) {
    return `${Math.round(num / 100000) / 10}M`;
  }
  return `${Math.round(num / 100) / 10}K`;
};

const MapChart = ({ setTooltipContent }) => {
  return (
    <>
      <ComposableMap height={450} data-tip="" projectionConfig={{ scale: 150 }}>
        <Geographies geography={geoUrl}>
          {({ geographies }) =>
            geographies.map((geo) => (
              <Geography
                key={geo.rsmKey}
                geography={geo}
                onMouseEnter={() => {
                  const { NAME, POP_EST } = geo.properties;
                  setTooltipContent(`${NAME} â€” ${rounded(POP_EST)}`);
                }}
                onMouseLeave={() => {
                  setTooltipContent('');
                }}
                style={{
                  default: {
                    fill: '#D6D6DA',
                    outline: 'none',
                    fontSize: 50,
                  },
                  hover: {
                    fill: '#F53',
                    outline: 'none',
                  },
                  pressed: {
                    fill: '#E42',
                    outline: 'none',
                  },
                }}
              />
            ))
          }
        </Geographies>
      </ComposableMap>
    </>
  );
};

export default memo(MapChart);
