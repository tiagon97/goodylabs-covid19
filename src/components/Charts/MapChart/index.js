import React, { memo } from 'react';
import { ComposableMap, Geographies, Geography } from 'react-simple-maps';
import PropTypes from 'prop-types';

import { Colors, Url } from '../../../constants';
import messages from './messages';
import CustomTooltip from './CustomTooltip';

const MapChart = ({ setTooltipContent, countriesData }) => (
  <>
    <ComposableMap height={420} data-tip="" projectionConfig={{ scale: 150 }}>
      <Geographies geography={Url.geoUrl}>
        {({ geographies }) =>
          geographies.map((geo) => (
            <Geography
              key={geo.rsmKey}
              geography={geo}
              onMouseEnter={() => {
                const { NAME, ISO_A2 } = geo.properties;
                const result = countriesData.filter((country) => country.CountryCode === ISO_A2);
                if (!result.length) {
                  setTooltipContent(<CustomTooltip header={NAME} />);
                } else {
                  setTooltipContent(
                    <CustomTooltip
                      header={NAME}
                      data={{
                        totalConfirmed: result[0].TotalConfirmed,
                        totalDeaths: result[0].TotalDeaths,
                        totalRecovered: result[0].TotalRecovered,
                      }}
                    />,
                  );
                }
              }}
              onMouseLeave={() => {
                setTooltipContent('');
              }}
              style={{
                default: {
                  fill: Colors.default,
                  outline: 'none',
                },
                hover: {
                  fill: Colors.hover,
                  outline: 'none',
                },
                pressed: {
                  fill: Colors.pressed,
                  outline: 'none',
                },
              }}
            />
          ))
        }
      </Geographies>
    </ComposableMap>
  </>
);

MapChart.propTypes = {
  setTooltipContent: PropTypes.func,
  countriesData: PropTypes.arrayOf(PropTypes.object).isRequired,
};

MapChart.defaultProps = {
  setTooltipContent: () => messages.noData,
};

export default memo(MapChart);
