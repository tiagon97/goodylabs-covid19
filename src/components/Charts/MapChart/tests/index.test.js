/**
 *
 * Tests for MapChart
 *
 */

import React from 'react';
import { render } from '@testing-library/react';

import MapChart from '../index';
import CustomTooltip from '../CustomTooltip';

describe('<MapChart />', () => {
  const defaultProps = {
    setTooltipContent: jest.fn(),
    countriesData: [
      {
        Country: 'Afghanistan',
        CountryCode: 'AF',
        Date: '2020-05-24T12:04:57Z',
        NewConfirmed: 782,
        NewDeaths: 11,
        NewRecovered: 44,
        Slug: 'afghanistan',
        TotalConfirmed: 9998,
        TotalDeaths: 216,
        TotalRecovered: 1040,
      },
      {
        Country: 'Albania',
        CountryCode: 'AL',
        Date: '2020-05-24T12:04:57Z',
        NewConfirmed: 8,
        NewDeaths: 0,
        NewRecovered: 6,
        Slug: 'albania',
        TotalConfirmed: 989,
        TotalDeaths: 31,
        TotalRecovered: 783,
      },
    ],
  };
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<MapChart {...defaultProps} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<MapChart {...defaultProps} />);
    expect(firstChild).toMatchSnapshot();
  });
});

describe('<CustomTooltip />', () => {
  const defaultProps = {
    header: 'Country name',
    data: {
      totalConfirmed: 100,
      totalDeaths: 10,
      totalRecovered: 40,
    },
  };
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<CustomTooltip {...defaultProps} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot with data', () => {
    const {
      container: { firstChild },
    } = render(<CustomTooltip {...defaultProps} />);
    expect(firstChild).toMatchSnapshot();
  });

  it('Should render and match the snapshot without data', () => {
    const {
      container: { firstChild },
    } = render(<CustomTooltip header="No data" />);
    expect(firstChild).toMatchSnapshot();
  });
});
