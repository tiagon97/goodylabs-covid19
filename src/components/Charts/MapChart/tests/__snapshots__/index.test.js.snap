// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomTooltip /> Should render and match the snapshot with data 1`] = `
<div
  class="sc-AxirZ bslVYb"
>
  <h4
    class="sc-AxjAm frcPRK"
  >
    Country name
  </h4>
  <div
    class="sc-AxiKw bOTXee"
  >
    <div
      class="sc-AxirZ sc-AxhCb bDKeGG"
    >
      <span
        class="sc-AxgMl jhmlNC"
        color="primary"
      >
        Total cases: 100
      </span>
      <span
        class="sc-AxgMl dYwBJV"
        color="secondary"
      >
        Total deaths: 10
      </span>
      <span
        class="sc-AxgMl imlJVq"
      >
        Total recovered: 40
      </span>
    </div>
  </div>
</div>
`;

exports[`<CustomTooltip /> Should render and match the snapshot without data 1`] = `
<div
  class="sc-AxirZ bslVYb"
>
  <h4
    class="sc-AxjAm frcPRK"
  >
    No data
  </h4>
  <div
    class="sc-AxiKw bOTXee"
  >
    No data available
  </div>
</div>
`;

exports[`<MapChart /> Should render and match the snapshot 1`] = `
<svg
  class="rsm-svg "
  data-tip=""
  viewBox="0 0 800 420"
>
  <g
    class="rsm-geographies "
  />
</svg>
`;
